### PR TITLE
Updates and cleanup for some configuration-related items (2.6.x)

### PR DESCRIPTION
**Pull request by JCgH4164838Gh792C124B5**
Updates and cleanup for some configuration-related items.
- Made several attributes final.
- Added limiter logic to the setting of certain configuration elements, and increased logging around them.
- Updated unit tests impacted by the limiter logic, including minor typo fixes and locale consistency.
- Minor cleanup while making the changes.

### CODE REVIEW

File: core/src/main/java/com/opensymphony/xwork2/config/providers/XWorkConfigurationProvider.java [hunk 1]
```diff
@@ -224,6 +224,7 @@ public void register(ContainerBuilder builder, LocatableProperties props)
         props.setProperty(StrutsConstants.STRUTS_ENABLE_OGNL_EXPRESSION_CACHE, Boolean.TRUE.toString());
         props.setProperty(StrutsConstants.STRUTS_ENABLE_OGNL_EVAL_EXPRESSION, Boolean.FALSE.toString());
         props.setProperty(StrutsConstants.STRUTS_CONFIGURATION_XML_RELOAD, Boolean.FALSE.toString());
+        props.setProperty(StrutsConstants.STRUTS_ALLOW_STATIC_METHOD_ACCESS, Boolean.FALSE.toString());
```
Comment (aleksandr-m) [inline core/src/main/java/com/opensymphony/xwork2/config/providers/XWorkConfigurationProvider.java]: Why? Is it not working?

File: core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java [hunk 2]
```diff
@@ -82,7 +86,18 @@ public void setXWorkConverter(XWorkConverter conv) {
 
     @Inject(StrutsConstants.STRUTS_DEVMODE)
     public void setDevMode(String mode) {
-        this.devMode = BooleanUtils.toBoolean(mode);
+        final boolean booleanMode = BooleanUtils.toBoolean(mode);
```
Comment (aleksandr-m) [inline core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: Why? If I remember correctly then injection can happen multiple times and some plugins, for example, can override already injected properties. That is how some things work (`disallowProxyMemberAccess` is one of this things).

File: core/src/main/resources/struts-default.xml [hunk 3]
```diff
@@ -61,6 +61,7 @@
                 freemarker.ext.rhino.,
                 sun.reflect.,
                 javassist.,
+                org.objectweb.asm.,
```
Comment (aleksandr-m) [inline core/src/main/resources/struts-default.xml]: This must be in separate commit / PR.

File: core/src/test/java/com/opensymphony/xwork2/ognl/OgnlUtilTest.java [hunk 4]
```diff
@@ -375,7 +381,7 @@ public void testSetPropertiesDate() {
         assertEquals(cal.getTime(), foo.getBirthday());
 
         //UK style test
-        cal = Calendar.getInstance();
+        cal = Calendar.getInstance(Locale.UK);
```
Comment (aleksandr-m) [inline core/src/test/java/com/opensymphony/xwork2/ognl/OgnlUtilTest.java]: Why?

File: core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java [hunk 5]
```diff
@@ -57,17 +57,21 @@
...
+    private boolean enableEvalExpressionValueSet;  // Defaults to false (not explicitly set yet)
 
     private Set<Class<?>> excludedClasses;
     private Set<Pattern> excludedPackageNamePatterns;
     private Set<String> excludedPackageNames;
 
     private Container container;
-    private boolean allowStaticMethodAccess;
-    private boolean disallowProxyMemberAccess;
+    private boolean allowStaticMethodAccess;            // Defaults to false
```
Comment (aleksandr-m) [inline core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: No need for the comments here.

File: core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java (hunk 2 above)
Comment (yasserzamani) [reply core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: @apache/struts-committers , I think here @JCgH4164838Gh792C124B5 is trying to prevent hacker from the ability of resetting those values that menaces overall security. Whether it works or breaks here, In general it seems a nice devised practice :+1: thank you!

Comment (aleksandr-m) [reply core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: @yasserzamani How it is helpful if basic stuff isn't working?

If this is really the case then they can be protected in some other way.

Comment (yasserzamani) [reply core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: Yeah basic stuff should work. I actually meant the idea itself - to prevent hacker from the ability of resetting those values that menaces overall security.

File: core/src/main/java/com/opensymphony/xwork2/config/providers/XWorkConfigurationProvider.java (hunk 1 above)
Comment (JCgH4164838Gh792C124B5) [reply core/src/main/java/com/opensymphony/xwork2/config/providers/XWorkConfigurationProvider.java]: Hello @aleksandr-m .
Are you asking why the new entry on line 227 ?  Simply to ensure the props includes a set property for that element, in line with the other significant ones already present.

File: core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java (hunk 2 above)
Comment (JCgH4164838Gh792C124B5) [reply core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: Hello Struts Team.

The primary intention for this PR was only to try and limit the ability to change certain elements (such as devMode), for which there didn't seem to be a good reason for them to change after the 1st (initial) configuration.  

It was more aimed at avoiding inadvertent/accidental state changes for important state than anything else.

None of the regression tests, nor interaction with the Showcase application encountered any issues so it seemed like a reasonable approach to take.

If there are certain properties (such as `disallowProxyMemberAccess`) for which subsequent changes are necessary/valid, then they could be omitted by this or another mechanism that attempts to limit such changes.

File: core/src/test/java/com/opensymphony/xwork2/ognl/OgnlUtilTest.java (hunk 4 above)
Comment (JCgH4164838Gh792C124B5) [reply core/src/test/java/com/opensymphony/xwork2/ognl/OgnlUtilTest.java]: Hello @aleksandr-m .

I found during debugging within an IDE that the `OgnlUtilTest` test `testSetPropertiesDate` was failing (even though it worked when run in the Maven build - on Windows in both cases).  It might be a combination of OS/environment conditions for the default Java locale - so I explicitly set the locale values (based on the existing comments) so that the test worked consistently both inside and outside the Maven build.

File: core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java (hunk 2 above)
Comment (JCgH4164838Gh792C124B5) [reply core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: Hello @aleksandr-m  and @yasserzamani .  

Do either of you think _any_ of the properties I attempted to apply change-limiters to could/should have the change-limiter applied to them ?

I could eliminate some or all of the change-limiters in a revision of this PR (and PR #291) that only retains the additional logging around property changes.  Please let me know what you think.

File: core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java (hunk 5 above)
Comment (aleksandr-m) [reply core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: Don't be. :) Unnecessary comments just clutter code and can take focus away from the important stuff.

File: core/src/test/java/com/opensymphony/xwork2/ognl/OgnlUtilTest.java (hunk 4 above)
Comment (aleksandr-m) [reply core/src/test/java/com/opensymphony/xwork2/ognl/OgnlUtilTest.java]: Ok.

File: core/src/main/java/com/opensymphony/xwork2/config/providers/XWorkConfigurationProvider.java (hunk 1 above)
Comment (aleksandr-m) [reply core/src/main/java/com/opensymphony/xwork2/config/providers/XWorkConfigurationProvider.java]: But there is more constants. Why to add especially this one?

File: core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java (hunk 2 above)
Comment (aleksandr-m) [reply core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: > Do either of you think any of the properties I attempted to apply change-limiters to could/should have the change-limiter applied to them ?

Well, currently I don't see what benefits does it bring.

File: core/src/main/java/com/opensymphony/xwork2/config/providers/XWorkConfigurationProvider.java (hunk 1 above)
Comment (JCgH4164838Gh792C124B5) [reply core/src/main/java/com/opensymphony/xwork2/config/providers/XWorkConfigurationProvider.java]: Properties like devMode, DMI, eval_expression are all "privilege-related" (and set to false), just like static method access.  Since the others were present, it seemed that static method access should be there too (for completeness, also set to false).
That was my reasoning anyway.  :)

File: core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java (hunk 2 above)
Comment (yasserzamani) [reply core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: Hi @JCgH4164838Gh792C124B5 , I like your idea but to be frank, I'm not able to have a vision on future side effects.

But I remembered one of my old ideas that not only has somehow same benefits but also is lower risky and is more readable: **Refactoring all methods annotated with `@com.opensymphony.xwork2.inject.Inject` from `public setXXX` to `private injectXXX`.** This preserves those secure against OGNL access. As a glimpse, with IntelliJ IDE's refactoring tools, I did it on `OgnlUtil.setDevMode` and worked  :)

File: core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java [hunk 6]
```diff
@@ -66,9 +76,14 @@ public void setAdditionalExcludePatterns(String excludePatterns) {
     @Inject(StrutsConstants.STRUTS_ENABLE_DYNAMIC_METHOD_INVOCATION)
     public void setDynamicMethodInvocation(String dmiValue) {
         if (!BooleanUtils.toBoolean(dmiValue)) {
-            LOG.debug("DMI is disabled, adding DMI related excluded patterns");
+            if (LOG.isDebugEnabled()) {
```
Comment (yasserzamani) [inline core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java]: Please revise all `if (LOG...` statements in this PR according to [this answer](https://stackoverflow.com/a/29561054/1362623).

File: core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java [hunk 7]
```diff
@@ -92,10 +107,17 @@ public void setEnableExpressionCache(String cache) {
 
     @Inject(value = StrutsConstants.STRUTS_ENABLE_OGNL_EVAL_EXPRESSION, required = false)
     public void setEnableEvalExpression(String evalExpression) {
-        enableEvalExpression = "true".equals(evalExpression);
-        if(enableEvalExpression){
-            LOG.warn("Enabling OGNL expression evaluation may introduce security risks " +
-                    "(see http://struts.apache.org/release/2.3.x/docs/s2-013.html for further details)");
+        if (!this.enableEvalExpressionValueSet) {
```
Comment (yasserzamani) [inline core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java]: Sorry you did a lot of work but please replace all change-limiters logic including their unit test with proposed solution above. Thanks in advance!

**Comment by JCgH4164838Gh792C124B5** [comment]
Hello @yasserzamani  and @aleksandr-m .

This PR has been updated with changes that hopefully addresses most of the concerns that were provided in your feedback.

The limiter logic was removed and instead of the full-refactoring of the injectable methods it only reduces the access level of most of the relevant methods to `protected` (improves access safety with less chance of impacting plugins or other elements negatively).  It's not the level of refactoring that @yasserzamani  suggested above, but is _a small step in that direction_ (although slightly less risky in terms of potential impacts).

Let me know if the changes in the PR look better now.

File: core/src/main/java/com/opensymphony/xwork2/security/DefaultAcceptedPatternsChecker.java [hunk 8]
```diff
@@ -70,7 +70,14 @@ public void setAcceptedPatterns(String[] additionalPatterns) {
     }
 
     public void setAcceptedPatterns(Set<String> patterns) {
-        LOG.trace("Sets accepted patterns [{}]", patterns);
+        if (acceptedPatterns == null) {
+            LOG.debug("Replacing accepted patterns [{}] with [{}], be aware that this affects all instances and safety of your application!",
```
Comment (yasserzamani) [inline core/src/main/java/com/opensymphony/xwork2/security/DefaultAcceptedPatternsChecker.java]: It should be `LOG.debug("Sets accepted patterns [{}]", patterns);`

File: core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java [hunk 9]
```diff
@@ -46,9 +46,15 @@ public DefaultExcludedPatternsChecker() {
...
-    public void setOverrideExcludePatterns(String excludePatterns) {
-        LOG.warn("Overriding excluded patterns [{}] with [{}], be aware that this affects all instances and safety of your application!",
-                    StrutsConstants.STRUTS_OVERRIDE_EXCLUDED_PATTERNS, excludePatterns);
+    protected void setOverrideExcludePatterns(String excludePatterns) {
+        if (excludedPatterns != null && excludedPatterns.size() > 0) {
+            LOG.warn("Overriding excluded patterns [{}] with [{}], be aware that this affects all instances and safety of your application!",
+                        excludedPatterns, excludePatterns);
+        }
+        else {
+            LOG.debug("Overriding excluded patterns [{}] with [{}], be aware that this affects all instances and safety of your application!",
```
Comment (yasserzamani) [inline core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java]: It should be `LOG.debug("Overriding excluded patterns with [{}]", excludePatterns);`

File: core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java [hunk 10]
```diff
@@ -80,7 +89,14 @@ public void setExcludedPatterns(String[] patterns) {
     }
 
     public void setExcludedPatterns(Set<String> patterns) {
-        LOG.trace("Sets excluded patterns [{}]", patterns);
+        if (excludedPatterns != null && excludedPatterns.size() > 0) {
+            LOG.warn("Replacing excluded patterns [{}] with [{}], be aware that this affects all instances and safety of your application!",
+                        excludedPatterns, patterns);
+        }
+        else {
+            LOG.debug("Replacing excluded patterns [{}] with [{}], be aware that this affects all instances and safety of your application!",
```
Comment (yasserzamani) [inline core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java]: It should be `LOG.debug("Sets excluded patterns [{}]", patterns);`

Comment (JCgH4164838Gh792C124B5) [reply core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java]: Hello @yasserzamani .
The two `debug` log statements have been updated in the most recent commit based on your comments/request.  I left in the extra warning text information but the wording and parameters of the start of the log output match what you requested.  :)

Comment (yasserzamani) [reply core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java]: Yes it's great thank you :)

@aleksandr-m , if you consent please approve your review, then we're OK to squash and merge this one :)

Comment (aleksandr-m) [reply core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java]: @yasserzamani Done. Squashing would be great. ;)
